src/chainstate/stacks/index/storage.rs

```diff
@@ -919,154 +646,133 @@ impl TrieFileStorage {
+        if !trie_sql::lock_bhh_for_extension(&mut self.db, bhh)? {
+            warn!("Block already extended: {}", &bhh);
+            return Err(Error::ExistsError);
+        }
 
-        TrieRAM::write_trie_file(&mut fd, &[], &[], identifier, bhh)?;
+        trace!("Extended from {} to {}", &self.cur_block, bhh);
 
-        fd.write_all(self.cur_block.as_bytes())
-            .map_err(|e| Error::IOError(e))?;
+        // write the new file out and add its parent
+        // TODO: I don't think this is necessary in the SQL impl,
+        //   so I've commented out -- but if it turns out we need to write
+        //   the parent block somewhere at this point, I've left this todo here so as not
+        //   to forget.
+        //  TrieRAM::write_trie_file(&mut fd, &[], &[], identifier, bhh)?;
+        //  fd.write_all(self.cur_block.as_bytes())
+        //    .map_err(|e| Error::IOError(e))?;
 
         // update internal structures
         self.cur_block = bhh.clone();
-        self.cur_block_fd = None;
+        self.cur_block_id = None;
 
         self.last_extended = Some((bhh.clone(), trie_buf));
 
-        trace!("Extended to {:?} in {:?}", &self.cur_block, &block_path);
+        Ok(())
+    }
+
+    // used for providing a option<block identifier> when re-opening a block --
+    //   because the previously open block may have been the last_extended block,
+    //   id may have been None.
+    pub fn open_block_maybe_id(&mut self, bhh: &BlockHeaderHash, id: Option<u32>) -> Result<(), Error> {
+        match id {
+            Some(id) => self.open_block_known_id(bhh, id),
+            None => self.open_block(bhh)
+        }
+    }
+
+    // used for providing a block identifier when opening a block -- usually used
+    //   when following a backptr, which stores the block identifier directly.
+    pub fn open_block_known_id(&mut self, bhh: &BlockHeaderHash, id: u32) -> Result<(), Error> {
+        if *bhh == self.cur_block && self.cur_block_id.is_some() {
+            // no-op
+            return Ok(())
+        }
+
+        if let Some((ref last_extended, _)) = self.last_extended {
+            if last_extended == bhh {
+                panic!("BUG: passed id of a currently building block");
+            }
+        }
+
+        // opening a different Trie than the one we're extending
+        self.cur_block_id = Some(id);
+        self.cur_block = bhh.clone();
+
         Ok(())
     }
 
     pub fn open_block(&mut self, bhh: &BlockHeaderHash) -> Result<(), Error> {
-        if *bhh == self.cur_block && self.cur_block_fd.is_some() {
+        if *bhh == self.cur_block && self.cur_block_id.is_some() {
             // no-op
             return Ok(())
         }
 
         let sentinel = TrieFileStorage::block_sentinel();
         if *bhh == sentinel {
             // just reset to newly opened state
-            self.cur_block_fd = None;
             self.cur_block = sentinel;
+            // did we write to the sentinel ?
```

Comment:
If we open the sentinel block, it would be because the caller is in the process of building up a TrieRAM (which may necessitate opening existing Tries on disk, and then switching back to the TrieRAM later).

Reply:
I think some of our tests abuse the sentinel, extending the trie to the sentinel, and writing to it. I think that's probably problematic in its own right (and think we may want to just use an Option value to indicate that there is no current block hash), but didn't want to deal with it in this PR to try to minimize the changes (and try not necessitate any changes to tests).

Reply:
Gotcha